New Translator Proxy

GitGab19 opened the pull request:
This PR replace the current Translator Proxy with a new one, which is able to:
- run in `Aggregated` or `Non-aggregated` mode
- ...

TODOs:
- Fix clippy
- Fix integration tests
- Squash commits into a single one (TBD)

GitGab19 commented:
This PR is now ready for review, and ready to be properly tested!

lucasbalieiro commented:
Nice! I'll start a long run testing session with my bitaxe.

plebhash commented:
> Mm that's strange, are you sure you were on the last commit?

it was @Shourya742 's machine

GitGab19 commented:
I've done changes to allow optional vardiff in tProxy.

But I tested it a lot with the Pool and it was working fine, so there's should be no difference with the JDC..

plebhash commented:
I spent a few hours today writing a new ITF module that integrates `minerd` in favor of `mining-device-sv1`, since `mining-device-sv1` was the main suspect from the investigations I did with @lucasbalieiro 

https://github.com/plebhash/stratum/tree/2025-08-30-minerd-itf

it was an interesting experiment, but unfortunately it did not fix the problems I reported on the message above

I used this ITF with `minerd` on `new-tproxy` + `new-jdc` branches and `jd_tproxy_integration` still hangs, so the mistery remains :(

GitGab19 commented:
Did you try to use the old tProxy? Or old JDC + new tProxy?

GitGab19 commented on roles/translator/Cargo.toml:
```diff
@@ -20,11 +20,17 @@ name = "translator_sv2"
+roles_logic_sv2 = { path = "../../protocols/v2/roles-logic-sv2" }
+stratum_translation = { path = "../roles-utils/stratum-translation" }
 serde = { version = "1.0.89", default-features = false, features = ["derive", "alloc"] }
 serde_json = { version = "1.0.64", default-features = false, features = ["alloc"] }
```
`stratum-translation` is used.

Were you referring to `roles_logic_sv2`?

GitGab19 commented on roles/translator/Cargo.toml:
```diff
@@ -20,11 +20,17 @@ name = "translator_sv2"
+stratum_translation = { path = "../roles-utils/stratum-translation" }
 serde = { version = "1.0.89", default-features = false, features = ["derive", "alloc"] }
 serde_json = { version = "1.0.64", default-features = false, features = ["alloc"] }
 futures = "0.3.25"
```
`stratum-translation` is used.

Were you referring to `roles_logic_sv2`?

GitGab19 commented:
This PR is rebased on top of https://github.com/stratum-mining/stratum/pull/1843, where all integration tests pass.

But I just noticed that currently there's another test which is failing on CI here: `jds_receive_solution_while_processing_declared_job_test`.

That test doesn't involve tProxy at all, so I really don't know how it can be failing here an not on https://github.com/stratum-mining/stratum/pull/1843.

cc @Shourya742 @plebhash

plebhash commented on roles/translator/src/lib/sv2/channel_manager/channel.rs:
```diff
@@ -0,0 +1,36 @@
+
+#[derive(Clone, Debug)]
+pub struct ChannelState {
+    pub upstream_sender: Sender<EitherFrame>,
+    pub upstream_receiver: Receiver<EitherFrame>,
```
do we really need `EitherFrame` here?

this is an `enum` that carries either a Sv2 message or a Handshake frame

plebhash commented on roles/translator/src/lib/sv2/upstream/upstream.rs:
```diff
@@ -0,0 +1,483 @@
+
+                    AnyMessage::Mining(_) => {
+                        // Forward mining message to channel manager
+                        let frame_to_forward = EitherFrame::Sv2(std_frame);
```
apparently this is related to the fact that since the `Sv2Frame` type carries a static lifetime, when we try to drop the channels inside `network_helpers_sv2`, the message prevents the channel from being dropped

plebhash commented:
`header_timestamp_value_assertion_in_new_extended_mining_job` is going to fail here, because new tProxy doesn't open a channel immediately with Pool

can be fixed with this:

```diff
diff --git a/test/integration-tests/tests/pool_integration.rs b/test/integration-tests/tests/pool_integration.rs
index abda15c6..88e86b80 100644
--- a/test/integration-tests/tests/pool_integration.rs
+++ b/test/integration-tests/tests/pool_integration.rs
@@ -104,7 +104,8 @@ async fn header_timestamp_value_assertion_in_new_extended_mining_job() {
         vec![],
         None,
     );
-    let _tproxy_addr = start_sv2_translator(pool_translator_sniffer_addr);
+    let (_tproxy, tproxy_addr) = start_sv2_translator(pool_translator_sniffer_addr);
+    start_mining_device_sv1(tproxy_addr, false, None);
 
     tp_pool_sniffer
         .wait_for_message_type(
```

plebhash commented:
`translate_sv1_to_sv2_successfully` is going to fail here, presumably because tProxy takes a few seconds to bootstrap and start sending messages, and this test is using the old syntax with macros `assert_*_message!` that expect the message to be immediately available when the assertion is called

it can be fixed with
```diff
diff --git a/test/integration-tests/tests/translator_integration.rs b/test/integration-tests/tests/translator_integration.rs
index f01cd955..2c38e13c 100644
--- a/test/integration-tests/tests/translator_integration.rs
+++ b/test/integration-tests/tests/translator_integration.rs
@@ -3,7 +3,6 @@ use integration_tests_sv2::{interceptor::MessageDirection, template_provider::Di
 use stratum_common::roles_logic_sv2::{
     common_messages_sv2::*,
     mining_sv2::*,
-    parsers_sv2::{AnyMessage, CommonMessages, Mining},
 };
 
 // This test runs an sv2 translator between an sv1 mining device and a pool. the connection between
@@ -22,26 +21,18 @@ async fn translate_sv1_to_sv2_successfully() {
     pool_translator_sniffer
         .wait_for_message_type(MessageDirection::ToUpstream, MESSAGE_TYPE_SETUP_CONNECTION)
         .await;
-    assert_common_message!(
-        &pool_translator_sniffer.next_message_from_downstream(),
-        SetupConnection
-    );
-    assert_common_message!(
-        &pool_translator_sniffer.next_message_from_upstream(),
-        SetupConnectionSuccess
-    );
-    assert_mining_message!(
-        &pool_translator_sniffer.next_message_from_downstream(),
-        OpenExtendedMiningChannel
-    );
-    assert_mining_message!(
-        &pool_translator_sniffer.next_message_from_upstream(),
-        OpenExtendedMiningChannelSuccess
-    );
-    assert_mining_message!(
-        &pool_translator_sniffer.next_message_from_upstream(),
-        NewExtendedMiningJob
-    );
+    pool_translator_sniffer
+        .wait_for_message_type(MessageDirection::ToDownstream, MESSAGE_TYPE_SETUP_CONNECTION_SUCCESS)
+        .await;
+    pool_translator_sniffer
+        .wait_for_message_type(MessageDirection::ToUpstream, MESSAGE_TYPE_OPEN_EXTENDED_MINING_CHANNEL)
+        .await;
+    pool_translator_sniffer
+        .wait_for_message_type(MessageDirection::ToDownstream, MESSAGE_TYPE_OPEN_EXTENDED_MINING_CHANNEL_SUCCESS)
+        .await;
+    pool_translator_sniffer
+        .wait_for_message_type(MessageDirection::ToDownstream, MESSAGE_TYPE_NEW_EXTENDED_MINING_JOB)
+        .await;
     pool_translator_sniffer
         .wait_for_message_type(
             MessageDirection::ToUpstream,
~
```

plebhash commented on roles/translator/config-examples/tproxy-config-hosted-pool-example.toml:
```diff
@@ -3,11 +3,6 @@
 # upstream_address = "18.196.32.109"
 # upstream_port = 3336
```
this info is outdated, shouldn't we remove it?

plebhash commented on roles/translator/config-examples/tproxy-config-local-jdc-example.toml:
```diff
@@ -3,11 +3,6 @@
 # upstream_address = "18.196.32.109"
```
ditto

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+            "Received OpenExtendedMiningChannelError from upstream with error code {}",
+            std::str::from_utf8(m.error_code.as_ref()).unwrap_or("unknown error code")
+        );
+        todo!("OpenExtendedMiningChannelError not handled yet");
```
I tried connecting to Braiins Pool, and forgot to set up a `user_identity` that matched the one I have created on the pool

that led to `OpenExtendedMiningChannel.Error` with `error_code = "unknown-user"`, which then panicked here

I guess we shouldn't really leave this as a `todo!`?

GitGab19 replied:
https://github.com/stratum-mining/stratum/issues/1870

GitGab19 commented on roles/translator/src/lib/utils.rs:
```diff
@@ -1,15 +1,296 @@
+    full_extranonce.extend_from_slice(extranonce1.as_slice());
+    full_extranonce.extend_from_slice(share.extra_nonce2.0.as_ref());
+
+    let share_version = share
+        .version_bits
+        .clone()
+        .map(|vb| vb.0)
+        .unwrap_or(job.version.0);
+    let mask = version_rolling_mask.unwrap_or(HexU32Be(0x1FFFE000_u32)).0;
+    let version = (job.version.0 & !mask) | (share_version & mask);
```
I took it from the old tProxy: https://github.com/stratum-mining/stratum/blob/d6a0a8783e24de857e7775f415fc5bacbedb5288/roles/translator/src/lib/proxy/bridge.rs#L338-L343

It seems that comes from this: https://github.com/slushpool/stratumprotocol/blob/master/stratum-extensions.mediawiki#changes-in-request-miningsubmit

GitGab19 commented on roles/translator/src/lib/sv1/sv1_server/sv1_server.rs:
```diff
@@ -0,0 +1,943 @@
+        let job_version = match message.job_version {
+            Some(version) => version,
+            None => {
+                warn!("Received share submission without valid job version, skipping");
```
@plebhash is this happening only on shares related to the very first job or it does impact every share?

Because I'm experiencing that only at the very beginning, but after the second job arrives, it doesn't happen anymore.

I will investigate it more asap.

plebhash replied:
only first job, after the second one arrives it goes away

but still, isn't it bad that those shares are being lost?

it seems that `d.last_job_version_field` is not initialized properly as soon as the first job arrives, only the second?

this is my first impression, but I still don't have the big picture

plebhash commented on roles/translator/src/lib/sv1/downstream/data.rs:
```diff
@@ -0,0 +1,88 @@
+    pub version_rolling_mask: Option<HexU32Be>,
+    pub version_rolling_min_bit: Option<HexU32Be>,
+    pub last_job_version_field: Option<u32>,
+    pub authorized_worker_names: Vec<String>,
```
why is this a `Vec`?

plebhash commented:
I launched tProxy in non-aggregated mode and connected multiple instances of `minerd`... I allowed them to run for a while and then killed all of them

Pool had some share submission logs with `sequence_number` over 1000

then I connected a new instance of `minerd` and observed Pool logs, where share submission had `sequence_number` above 1000

this seems wrong, and it's behaving as if this was aggregated mode

on non-aggregated mode, each channel should submit shares while keeping `sequence_number` independent from other channels

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        &mut self,
+        _m: roles_logic_sv2::mining_sv2::SetCustomMiningJobSuccess,
+    ) -> Result<(), HandlerError> {
+        unreachable!("Cannot process SetCustomMiningJobSuccess since Translator Proxy does not support custom mining jobs")
```
yeah overall I don't think we should panic after receiving unexpected messages

plebhash commented on roles/translator/src/lib/config.rs:
```diff
@@ -137,6 +127,9 @@ pub struct DownstreamDifficultyConfig {
+    pub enable_vardiff: bool,
     /// The number of shares submitted since the last difficulty update.
     #[serde(default = "u32::default")]
     pub submits_since_last_update: u32,
```
`submits_since_last_update` and `timestamp_of_last_update` are never used for anything

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: roles_logic_sv2::mining_sv2::OpenMiningChannelError<'_>,
+    ) -> Result<(), HandlerError> {
+        error!(
+            "Received OpenExtendedMiningChannelError from upstream with error code {}",
+            std::str::from_utf8(m.error_code.as_ref()).unwrap_or("unknown error code")
```
here, we would only be showing error code, but not other relevant message fields

we have `Display` trait implementations for all messages now, we should leverage them
```suggestion
            "Received: {}", m
```

plebhash replied:
and the `debug!` print below could be removed

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: roles_logic_sv2::mining_sv2::CloseChannel<'_>,
+    ) -> Result<(), HandlerError> {
+        info!(
+            "Received CloseChannel from upstream for channel id: {}",
+            m.channel_id
```
here, we would only be showing`channel_id`, but not other relevant message fields

we have `Display` trait implementations for all messages now, we should leverage them
```suggestion
            "Received: {}", m
```

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        &mut self,
+        _m: roles_logic_sv2::mining_sv2::SetExtranoncePrefix<'_>,
+    ) -> Result<(), HandlerError> {
+        unreachable!("Cannot process SetExtranoncePrefix since set_extranonce is not supported for majority of sv1 clients");
```
yeah I'm not sure panic is the best option here

one simple choice now is to log a warning and do nothing

a better choice (that would probably take more time/effort to implement) would be to disconnect all Sv1 workers and make sure they're all following the new `extranonce_prefix` when they reconnect

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: roles_logic_sv2::mining_sv2::SubmitSharesSuccess,
+    ) -> Result<(), HandlerError> {
+        info!(
+            "Received SubmitSharesSuccess from upstream for channel id: {} ✅",
+            m.channel_id
```
here, we would only be showing `channel_id`, but not other relevant message fields

we have `Display` trait implementations for all messages now, we should leverage them
```suggestion
            "Received: {}", m
```

plebhash replied:
and the `debug!` print below could be removed

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: roles_logic_sv2::mining_sv2::SubmitSharesError<'_>,
+    ) -> Result<(), HandlerError> {
+        warn!(
+            "Received SubmitSharesError from upstream for channel id: {} ❌",
+            m.channel_id
```
ditto
```suggestion
            "Received: {}", m
```

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: NewExtendedMiningJob<'_>,
+    ) -> Result<(), HandlerError> {
+        info!(
+            "Received NewExtendedMiningJob from upstream for channel id: {}",
+            m.channel_id
```
ditto
```suggestion
            "Received: {}", m
```

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        let m_static = m.clone().into_static();
+        _ = self.channel_manager_data.safe_lock(|channel_manager_data| {
+            info!(
+                "Received SetNewPrevHash from upstream for channel id: {}",
+                m.channel_id
```
ditto
```suggestion
            "Received: {}", m
```

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+
+    async fn handle_set_target(&mut self, m: SetTarget<'_>) -> Result<(), HandlerError> {
+        info!(
+            "Received SetTarget from upstream for channel id: {}",
+            m.channel_id
```
ditto
```suggestion
            "Received: {}", m
```

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+            let (user_identity, nominal_hashrate, downstream_extranonce_len) = channel_manager_data
+            .pending_channels
+            .remove(&m.request_id)
+            .unwrap_or_else(|| ("unknown".to_string(), 100000.0, 0_usize));
```
what's the rationale behind these "default" values?

plebhash commented on roles/translator/src/lib/sv2/channel_manager/message_handler.rs:
```diff
@@ -0,0 +1,426 @@
+        m: roles_logic_sv2::mining_sv2::UpdateChannelError<'_>,
+    ) -> Result<(), HandlerError> {
+        error!(
+            "Received UpdateChannelError from upstream with error code {}",
+            std::str::from_utf8(m.error_code.as_ref()).unwrap_or("unknown error code")
```
and the `debug!` print below could be removed

GitGab19 commented on roles/translator/src/lib/sv1/downstream/data.rs:
```diff
@@ -0,0 +1,88 @@
+    pub version_rolling_min_bit: Option<HexU32Be>,
+    pub last_job_version_field: Option<u32>,
+    pub authorized_worker_names: Vec<String>,
+    pub is_authorized: RefCell<bool>,
```
It's something coming from some old iteration on this, removing it.

GitGab19 commented on roles/translator/src/lib/sv1/sv1_server/sv1_server.rs:
```diff
@@ -0,0 +1,943 @@
+        // Spawn vardiff loop only if enabled
+        if self.config.downstream_difficulty_config.enable_vardiff {
+            info!("Variable difficulty adjustment enabled - starting vardiff loop");
+            task_manager.spawn(DifficultyManager::spawn_vardiff_loop(
+                self.sv1_server_data.clone(),
+                self.sv1_server_channel_state.channel_manager_sender.clone(),
+                self.sv1_server_channel_state
+                    .sv1_server_to_downstream_sender
+                    .clone(),
+                self.shares_per_minute,
+                self.config.aggregate_channels,
+                notify_shutdown.subscribe(),
+                shutdown_complete_tx_main_clone.clone(),
```
As agreed with @Shourya742 , he's gonna tackle this in a follow-up PR, since from initial experiments this was introducing hanging tests in tProxy.

GitGab19 replied:
https://github.com/stratum-mining/stratum/issues/1871

GitGab19 commented on roles/translator/src/lib/sv2/upstream/channel.rs:
```diff
@@ -0,0 +1,41 @@
+    /// Sender for the SV2 Upstream role
+    pub upstream_sender: Sender<EitherFrame>,
+    /// Sender for the ChannelManager thread
+    pub channel_manager_sender: Sender<EitherFrame>,
```
https://github.com/stratum-mining/stratum/pull/1791#discussion_r2346811822

GitGab19 replied:
https://github.com/stratum-mining/stratum/issues/1872

GitGab19 commented on roles/translator/src/lib/sv2/upstream/upstream.rs:
```diff
@@ -0,0 +1,483 @@
+                                // Ignore other shutdown variants for upstream
+                            }
+                            Err(e) => {
+                                // Handle channel lag gracefully - don't shutdown on lag errors
+                                if let tokio::sync::broadcast::error::RecvError::Lagged(_) = e {
+                                    warn!("Upstream: broadcast channel lagged, continuing: {e}");
+                                } else {
+                                    error!("Upstream: failed to receive shutdown signal: {e}");
+                                    break;
+                                }
```
Removed.

GitGab19 commented on roles/translator/src/lib/sv1/downstream/message_handler.rs:
```diff
@@ -0,0 +1,155 @@
+};
+
+// Implements `IsServer` for `Downstream` to handle the SV1 messages.
+impl IsServer<'static> for DownstreamData {
```
> also, shouldn't `IsServer` trait be implemented for `Sv1Server` struct?
> 
> this is not based on a deep/comprehensive analysis of the entire codebase, I'm just saying it from an instinctual level

I have to think more about this, but since the beginning it was easier to implement that on the `Downstream`.

Consider that the `Sv1Server` maintains an Hashmap of `Downstream`, so they are strictly linked together.

GitGab19 commented on roles/translator/src/lib/sv1/downstream/data.rs:
```diff
@@ -0,0 +1,88 @@
+    // Flag to track if SV1 handshake is complete (subscribe + authorize)
+    pub sv1_handshake_complete: bool,
+    // Stores pending shares to be sent to the sv1_server
+    pub pending_share: RefCell<Option<SubmitShareWithChannelId>>,
```
The problem here is the `IsServer` trait, which defined the `handle_submit` method with `&self`. 

So we cannot mutate anything inside it, that's why I used `RefCell`.

The alternative could be to change the trait, but I didn't want to break the crate just for this.

GitGab19 commented on roles/translator/src/lib/sv2/upstream/message_handler.rs:
```diff
@@ -0,0 +1,54 @@
+            msg.error_code, msg.flags
+        );
+
+        todo!()
```
Here we should probably trigger a fallback, so I think this is strictly related to this comment: https://github.com/stratum-mining/stratum/pull/1791#discussion_r2346795935

GitGab19 replied:
https://github.com/stratum-mining/stratum/issues/1870

GitGab19 commented on roles/translator/src/lib/sv1/downstream/downstream.rs:
```diff
@@ -0,0 +1,477 @@
+            if let Err(e) = self
+                .downstream_channel_state
+                .sv1_server_sender
+                .try_send(DownstreamMessages::SubmitShares(share))
```
Changed into `send()`

GitGab19 commented on roles/translator/src/lib/sv2/upstream/channel.rs:
```diff
@@ -0,0 +1,41 @@
+    /// Sender for the ChannelManager thread
+    pub channel_manager_sender: Sender<EitherFrame>,
+    /// Receiver for the ChannelManager thread
+    pub channel_manager_receiver: Receiver<EitherFrame>,
```
https://github.com/stratum-mining/stratum/issues/1872

lucasbalieiro commented:
I’ve been testing this over the past few days and haven’t found any unexpected behavior.

I was able to mine blocks with both aggregated and non-aggregated mode, with multiple clients. I also tested the translation using `minerd` and a `bitaxe`, and everything worked as expected.

So far, the happy path scenario looks solid.

plebhash commented:
CI is broken here

by looking at the error logs, it seems that's because of #1884, which changed how some `channels_sv2` APIs should be used by JDC

as far as I remember, none of those changes should affect new tProxy... CI is only broken here because we're rebasing on top of #1843

plebhash commented:
I'm seeing some weird behavior with new tProxy in aggregated mode connected to JDC at a7a6bdd3e41eb093ae99f29677669e7247615213

the first weird behavior is `minerd` being thrown into a reconnection loop, presumably due to `OpenExtendedMiningChannel.Success.extranonce_size = 20`
```
2025-09-20T01:49:22.919337Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: OpenExtendedMiningChannelSuccess(request_id: 2, channel_id: 2, target: U256(000001ad7f26db37886a40019c4ccf8cf08f33fef18bcabc3d8bed7829fc5ff0), extranonce_size: 20, extranonce_prefix: B032(000100000000000100000002)), user_identity: your_username_here.miner2, nominal_hashrate: 1000000
2025-09-20T01:49:22.919412Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: NewExtendedMiningJob(channel_id: 2, job_id: 1, min_ntime: Sv2Option(None), version: 536870912, version_rolling_allowed: true, merkle_path: Seq0255<len=0: [], coinbase_tx_prefix: B064K(02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff3b0305b60000142f2f5376324d696e65725369676e61747572652f20), coinbase_tx_suffix: B064K(feffffff0200f2052a01000000160014ebe1b7dcc293ccaa0ee743a86f89df8258c208fc0000000000000000266a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf904b60000))
2025-09-20T01:49:22.919463Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: SetNewPrevHash(channel_id=2, job_id=1, prev_hash=U256(00000000000241de02f508142c85839b582cb0c8698cbe575c43b736cfa9ccfa), min_ntime=1758332954, nbits=453205760)
2025-09-20T01:49:24.366055Z ERROR translator_sv2::sv1::downstream::downstream: Down: Error handling downstream message: InvalidSubmission
2025-09-20T01:49:24.366093Z ERROR translator_sv2::sv1::downstream::downstream: Downstream 2: error in downstream message handler: SV1Error
2025-09-20T01:49:24.366100Z ERROR translator_sv2::status: Error in Downstream { downstream_id: 2, tx: Sender { .. } }: SV1Error
2025-09-20T01:49:24.366107Z  WARN translator_sv2::status: Downstream [2] shutting down due to error: SV1Error
2025-09-20T01:49:24.366117Z  WARN translator_sv2::sv1::downstream::downstream: Downstream 2: unified task shutting down
2025-09-20T01:49:24.366175Z  WARN translator_sv2: Downstream 2 disconnected — notifying SV1 server.
2025-09-20T01:49:24.366252Z  INFO translator_sv2::sv1::sv1_server::sv1_server: 🔌 Downstream: 2 disconnected and removed from sv1 server downstreams
2025-09-20T01:49:24.366264Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Sending close channel message: 2 for downstream: 2
2025-09-20T01:49:24.367010Z  INFO translator_sv2::sv1::sv1_server::sv1_server: New SV1 downstream connection from 127.0.0.1:50779
2025-09-20T01:49:24.367033Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Downstream 3 registered successfully (SV2 channel will be opened on first SV1 message)
2025-09-20T01:49:24.367095Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.subscribe from Sv1 downstream
2025-09-20T01:49:24.367249Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.authorize from Sv1 downstream
2025-09-20T01:49:24.367502Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: OpenExtendedMiningChannelSuccess(request_id: 3, channel_id: 3, target: U256(000001ad7f26db37886a40019c4ccf8cf08f33fef18bcabc3d8bed7829fc5ff0), extranonce_size: 20, extranonce_prefix: B032(000100000000000100000003)), user_identity: your_username_here.miner3, nominal_hashrate: 1000000
2025-09-20T01:49:24.367547Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: NewExtendedMiningJob(channel_id: 3, job_id: 1, min_ntime: Sv2Option(None), version: 536870912, version_rolling_allowed: true, merkle_path: Seq0255<len=0: [], coinbase_tx_prefix: B064K(02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff3b0305b60000142f2f5376324d696e65725369676e61747572652f20), coinbase_tx_suffix: B064K(feffffff0200f2052a01000000160014ebe1b7dcc293ccaa0ee743a86f89df8258c208fc0000000000000000266a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf904b60000))
2025-09-20T01:49:24.367595Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: SetNewPrevHash(channel_id=3, job_id=1, prev_hash=U256(00000000000241de02f508142c85839b582cb0c8698cbe575c43b736cfa9ccfa), min_ntime=1758332954, nbits=453205760)
2025-09-20T01:49:24.984659Z ERROR translator_sv2::sv1::downstream::downstream: Error receiving downstream message: RecvError
2025-09-20T01:49:24.984668Z ERROR translator_sv2::sv1::downstream::downstream: Downstream 3: error in downstream message handler: ChannelErrorReceiver(RecvError)
2025-09-20T01:49:24.984672Z ERROR translator_sv2::status: Error in Downstream { downstream_id: 3, tx: Sender { .. } }: ChannelErrorReceiver(RecvError)
2025-09-20T01:49:24.984677Z  WARN translator_sv2::status: Downstream [3] shutting down due to error: ChannelErrorReceiver(RecvError)
2025-09-20T01:49:24.984682Z  WARN translator_sv2::sv1::downstream::downstream: Downstream 3: unified task shutting down
2025-09-20T01:49:24.984689Z  WARN translator_sv2: Downstream 3 disconnected — notifying SV1 server.
2025-09-20T01:49:24.984731Z  INFO translator_sv2::sv1::sv1_server::sv1_server: 🔌 Downstream: 3 disconnected and removed from sv1 server downstreams
2025-09-20T01:49:24.984741Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Sending close channel message: 3 for downstream: 3
```

doing this patch to Pool made the error above go away:
```diff
diff --git a/roles/pool/src/lib/mining_pool/mod.rs b/roles/pool/src/lib/mining_pool/mod.rs
index bd4f04fd..8ba98764 100644
--- a/roles/pool/src/lib/mining_pool/mod.rs
+++ b/roles/pool/src/lib/mining_pool/mod.rs
@@ -983,7 +983,7 @@ impl Pool {
         // range_1 is used for dynamically allocating extranonce_prefix across different channels
         // from these 8 bytes, the first 2 bytes are statically defined by static_prefix
         let range_1_start = 0;
-        let range_1_end = 8;
+        let range_1_end = 24;
 
         // range_0 is not used here
         let range_0 = std::ops::Range {
```

but then, tProxy rejects all shares from CPU miner:
```
2025-09-20T01:53:32.188236Z  INFO translator_sv2: Starting Translator Proxy...
2025-09-20T01:53:32.188343Z  INFO translator_sv2::sv2::upstream::upstream: Trying to connect to upstream 0 at 127.0.0.1:34265
2025-09-20T01:53:32.188576Z  INFO translator_sv2::sv2::upstream::upstream: Connected to upstream at 127.0.0.1:34265 (attempt 1/3)
2025-09-20T01:53:32.190013Z  INFO translator_sv2::sv2::upstream::message_handler: Received: SetupConnectionSuccess(used_version: 2, flags: 4)
2025-09-20T01:53:32.190022Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Starting SV1 server on 0.0.0.0:34255
2025-09-20T01:53:32.190060Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Variable difficulty adjustment disabled - upstream will manage difficulty, SV1 server will forward SetTarget messages to downstreams
2025-09-20T01:53:32.190109Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Translator Proxy: listening on 0.0.0.0:34255
2025-09-20T01:53:42.813069Z  INFO translator_sv2::sv1::sv1_server::sv1_server: New SV1 downstream connection from 127.0.0.1:51198
2025-09-20T01:53:42.813116Z  INFO translator_sv2::sv1::sv1_server::sv1_server: Downstream 1 registered successfully (SV2 channel will be opened on first SV1 message)
2025-09-20T01:53:42.813278Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.subscribe from Sv1 downstream
2025-09-20T01:53:42.813495Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.authorize from Sv1 downstream
2025-09-20T01:53:42.814195Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: OpenExtendedMiningChannelSuccess(request_id: 1, channel_id: 1, target: U256(000001ad7f26db37886a40019c4ccf8cf08f33fef18bcabc3d8bed7829fc5ff0), extranonce_size: 4, extranonce_prefix: B032(00010000000000000000000000000000000000000000000100000001)), user_identity: your_username_here.miner1, nominal_hashrate: 1000000
2025-09-20T01:53:42.814273Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: NewExtendedMiningJob(channel_id: 1, job_id: 1, min_ntime: Sv2Option(None), version: 536870912, version_rolling_allowed: true, merkle_path: Seq0255<len=0: [], coinbase_tx_prefix: B064K(02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff3b0305b60000142f2f5376324d696e65725369676e61747572652f20), coinbase_tx_suffix: B064K(feffffff0200f2052a01000000160014ebe1b7dcc293ccaa0ee743a86f89df8258c208fc0000000000000000266a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf904b60000))
2025-09-20T01:53:42.814331Z  INFO translator_sv2::sv2::channel_manager::message_handler: Received: SetNewPrevHash(channel_id=1, job_id=1, prev_hash=U256(00000000000241de02f508142c85839b582cb0c8698cbe575c43b736cfa9ccfa), min_ntime=1758333202, nbits=453205760)
2025-09-20T01:53:44.214915Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:44.215013Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:44.447080Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:44.447151Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:44.820507Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:44.820556Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:44.999811Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:44.999901Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:46.689582Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:46.689692Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:46.973364Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:46.973433Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:47.054939Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:47.055017Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:48.096447Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
2025-09-20T01:53:48.096524Z ERROR translator_sv2::sv1::downstream::message_handler: Invalid share for channel id: 1
2025-09-20T01:53:48.449455Z  INFO translator_sv2::sv1::downstream::message_handler: Received mining.submit from Sv1 downstream for channel id: 1
```

---

@lucasbalieiro can you try to reproduce?

GitGab19 commented:
@plebhash your issues might have probably due to a recent change which I reverted in https://github.com/stratum-mining/stratum/pull/1791/commits/fc88bbd7f778c21d083b33b14d0455d41208214f.

I tried to make tProxy open a channel only after receiving the first SV1 message, but I guess I broke something, so I reverted it now.

I'm running tProxy with JDC, both aggregated and non-aggregated, and now everything is working fine again!
